Mcm 187

prashmigit opened the pull request:
<!--
Is there any breaking changes?  If so this is a major release, make sure '#major' is in at least one
commit message to get CI to bump the major.  This will prevent automatic down stream dependency
bumping / consuming.  For more information about semantic versioning see: https://semver.org/


Suggested PR template: Fill/delete/add sections as needed. Optionally delete any commented block.
-->
What
----
<!--
Briefly describe **what** you have changed and **why**.
Optionally include implementation strategy.
-->

References
----------
<!--
Copy&paste links: to Jira ticket, other PRs, issues, Slack conversations...
For code bumps: link to PR, tag or GitHub `/compare/master...master`
-->

Test&Review
------------

<!--
Has it been tested? how?
Copy&paste any handy instructions, steps or requirements that can save time to the reviewer or any reader.
-->

<!--
Open questions / Follow ups
--------------------------
<!--
Optional: anything open to discussion for the reviewer, out of scope, or follow ups.
-->

<!--
Review stakeholders
------------------
<!--
Optional: mention stakeholders or if special context that is required to review.
-->

dguy commented on upgrader/src/main/java/io/confluent/cloud/roller/rollexecution/RollTaskExecutor.java:
```diff
@@ -1,12 +1,112 @@
 
-    // triggerInternalClusterSanityRoll from the RollTask api
+    @Autowired
+    private RollController rollController;
```
See above

prashmigit replied:
addressed

dguy commented on upgrader/src/main/java/io/confluent/cloud/roller/rollexecution/chunking/Chunker.java:
```diff
@@ -0,0 +1,28 @@
+    public Chunker getChunkerFactory() {
+
+        if (chunkingStrategy.equalsIgnoreCase(CHUNKING_STRATEGY.DEFAULT.name())) {
+            return new DefaultChunker();
```
A class should not know about its subclasses.
Also, i see no need for inheritance - this should be an interface

prashmigit replied:
Good catch! thanks..  I forgot to complete my changes here. Was meant to be converted into a ChunkerFactory.
I will eventually move that into @Profile driven Chunker based on the environment.

prashmigit replied:
Good catch! thanks.. I forgot to complete my changes here. Was meant to be converted into a ChunkerFactory.
I will eventually move that into @Profile driven Chunker based on the environment.
Pushing the patch

prashmigit replied:
done, with Profiles in the latest patch

dguy commented on upgrader/src/main/java/io/confluent/cloud/roller/rolltask/RollTask.java:
```diff
@@ -1,5 +1,49 @@
 
+public class RollTask implements ChunkableTask {
+
+    private long rollId;
```
`final` for fields

prashmigit replied:
changed

dguy commented on upgrader/src/main/java/io/confluent/cloud/roller/rolltask/RollTask.java:
```diff
@@ -1,5 +1,49 @@
+    // subsequent Roll requests - this will avoid too many queries during time of execution
+    private int weight;
+
+    public RollTask(long rollId, String physicalClusterId, String clusterType) {
```
Don't know how this would pass checkstyle as the params should be `final`

prashmigit replied:
sure

prashmigit commented on upgrader/src/main/java/io/confluent/cloud/roller/roll/RollController.java:
```diff
@@ -1,22 +1,204 @@
+ */
 public class RollController {
 
+    @Autowired
```
I agree that putting all of them in constructors is one way of doing it. However, all of these are @Component annotated. They are managed singletons and are meant to be stateless.

dguy replied:
I'm not bothered about them being managed. I want to be able to create them without spring and all fields that can be `final` should be. It is just good practice..

prashmigit replied:
@dguy, This pattern is pretty widely used. It is easily unit testable as well with `Mockito` - I do not believe that while writing unit tests we should be burdened with constructing the full set of dependency objects for the class being tested. Any higher level of tests like integration and end to end tests can be written as well - `Autowired` does not impede testability. That being said, to get this going, I have switched to using parameterized constructors with finals. Since each dependency itself is already a Component my requirement is still being met.

prashmigit commented on upgrader/src/main/java/io/confluent/cloud/roller/roll/RollState.java:
```diff
@@ -0,0 +1,116 @@
+        return status;
+    }
+
+    public static class RollStateBuilder {
```
Seems this is mainly about https://howtodoinjava.com/design-patterns/creational/builder-pattern-in-java/ [above] vs https://dzone.com/articles/the-builder-pattern-for-class-with-many-constructo. 

The one I have used above is slightly more leaner - as lean as Builders can be, I suppose. None of the fields can be changed since all are `private` and *no public setters* are provided. The only control here is via the builder which creates a new RollState every time, making the RollState object immutable.
Please let me know in case I am missing something here.

dguy replied:
It isn't immutable as you can use the `RollStateBuilder` to indirectly change the state of `RollState`

prashmigit replied:
I see what you mean - fixed it.

prashmigit commented:
@dguy , thanks for taking a look. Yes tests are missing as called out in the PR notes as well. I would normally put tests as I go, but the circumstances in which I had to keep making progress made it harder than necessary. However, all of that is included in the Jira whcih completes only after all the open items are complete. I have addressed your comments.

dguy commented:
@prashmigit tests are an essential part of any PR. If the PR is too big, then break it up someway where possible, but please don't get into the habit of putting tests into another PR. It rarely works out well. 
For the record - i generally won't approve any PRs (with code) that don't have tests.

prashmigit commented on upgrader/src/main/java/io/confluent/cloud/roller/roll/RollState.java:
```diff
@@ -0,0 +1,116 @@
+
+public class RollState {
+    @JsonProperty("rollId")
+    private long rollId;
```
changed.

prashmigit commented:
> @prashmigit tests are an essential part of any PR. If the PR is too big, then break it up someway where possible, but please don't get into the habit of putting tests into another PR. It rarely works out well.
> For the record - i generally won't approve any PRs (with code) that don't have tests.

@dguy, Unit tests are now included.
